site/source/docs/porting/exceptions.rst

```diff
@@ -32,19 +23,53 @@ For example, build the following program with ``emcc -fexceptions``:
       return 0;
     }
 
-Running it will print ``throw...`` and then ``catch!``.
+The first ``throw`` will abort the program and you'll see something like this in the output:
+
+.. code-block:: text
+
+  throw...  
+  exception thrown: 5246024 - Exception catching is disabled, this exception cannot be caught. Compile with -s NO_DISABLE_EXCEPTION_CATCHING or -s EXCEPTION_CATCHING_ALLOWED=[..] to catch.
+
+If you want to opt-in, you have two following options.
+
+JavaScript-based exception support
+##################################
+
+First, you can enable exceptions via Emscripten's JavaScript-based support.
+
+To enable it, pass ``-fexceptions`` at both compile time and link time.
+
+When you rebuild the example above with this flag, the output will change to:
+
+.. code-block:: text
+
+  throw...
+  catch!
+
+Note that this option has relatively high overhead, but it will work on all JavaScript
+engines with WebAssembly support. You can reduce the overhead by specifying a
+list of allowed functions in which exceptions are enabled, see the
+``EXCEPTION_CATCHING_ALLOWED`` setting.
+
+WebAssembly exception handling proposal
+#######################################
+
+Alternatively, you can opt-in to the `native WebAssembly exception handling
+<https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md>`_.
+
+To enable it, pass ``-fwasm-exceptions`` at both compile time and link time.
+
+Rebuilding the example with this flag will result in the same output as with
+``-fexceptions`` above:
 
-Overhead
-########
+.. code-block:: text
 
-Until wasm gets direct support for exceptions, Emscripten supports them using
-JavaScript for throwing and catching. As a result, every time we might need to
-catch an exception we call out into JavaScript to a try-catch there. This
-increases code size and adds significant overhead.
+  throw...
+  catch!
 
-You can reduce the overhead by specifying a list of allowed functions in
-which exceptions are enabled, see the ``EXCEPTION_CATCHING_ALLOWED`` setting.
+This option leverages a new feature that brings built-in instructions for
+throwing and catching exceptions to WebAssembly.
 
-In the future direct support for exceptions in wasm will avoid that overhead.
-Usage will be as mentioned earlier, but with ``-fwasm-exceptions`` instead of
-``-fexceptions``.
+As a result, it can reduce amount of custom glue and performance overhead
```

Comment:
"custom glue" might not really be meaningful to a user. Maybe "reduce calls between wasm and JS" or something like that?

Reply:
I was rather trying to convey the code size reductions here, but not sure how to do that best. "reduce amount of Emscripten-generated JavaScript"? That would be a bit more verbose but maybe clearer than "glue"?

Reply:
I'm not sure the reduction in the glue size is important enough that it warrants a mention here. As we've seen, it involves small decrease in glue size and small increase in C library size but both of them are not really the point; the user code transformation EH handling incurs is a lot smaller in Wasm EH compared to the old Emscripten EH, and that's the main reason there are code size savings, so I agree w/ @dschuff that we don't need to mention this.

Reply:
> I agree w/ @dschuff that we don't need to mention this

I don't think that's what he's saying, he just asked to clarify it?

> the user code transformation EH handling incurs is a lot smaller in Wasm EH compared to the old Emscripten EH, and that's the main reason there are code size savings

Right, which is why I do want to mention them as they feel like also an important benefit.

Reply:
I think what @dschuff said is custom glue thing is too much of detail than what users would want to know, and what I said is, on top of that, that's not really the main source of code size savings anyway. So yeah, we have different reasons but maybe the same conclusions (to omit that part)? 😅 How about just simply saying
> As a result, it can reduce code size and performance overhead compared to the JavaScript-based implementation.

Reply:
> I think what @dschuff said is custom glue thing is too much of detail

Oh, I interpreted it as the opposite - that "custom glue" is too obscure for the user, aka it's not enough detail.

> As a result, it can reduce code size and performance overhead compared to the JavaScript-based implementation.

Sure, makes sense.

Reply:
Ah, I don't think I actually meant that "custom glue" was either not enough detail or too much detail, just that a user wouldn't know what it meant :)
Then I was thinking that the real benefit was reducing calls through JS (rather than size), but I guess that's covered under "performance overhead", so my suggestion wasn't great :)
Anyway, I like @aheejin's version best.

Reply:
Done